File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public ListNode RemoveNthFromEnd(ListNode head, int n) {
        var dummy = new ListNode(0, head);
        var f = dummy;
        var s = dummy;

        for(var i = 0; i <= n; i++)
            f = f.next;
        
        while(f != null)
        {
            f = f.next;
            s = s.next;
        }

        s.next = s.next.next;

        return dummy.next;
    }
}

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        group_prev = dummy

        while True:
            kth = self.getKth(group_prev, k)
            if not kth:
                break
            group_next = kth.next

            prev, curr = kth.next, group_prev.next
            while curr != group_next:
                t = curr.next
                curr.next = prev
                prev = curr
                curr = t
            
            tmp = group_prev.next
            group_prev.next = kth
            group_prev = tmp
        
        return dummy.next



    def getKth(self, curr, k):
        while k > 0 and curr:
            k -= 1
            curr = curr.next
        return curr

                



        

